src/useBluetoothNotifications.tsx

```diff
@@ -0,0 +1,188 @@
+import {
+  BluetoothNotificationsHookOptions,
+  BluetoothConnections,
+  BluetoothNotificationsStatus,
+  BluetoothEvent,
+  USER_CANCEL_ERROR_CODE,
+} from "./types/Bluetooth";
+import { useCallback, useState, useEffect, useMemo } from "react";
+import { getFloatValue } from "./utils/getFloatValue";
+
+export const defaultHookOptions: BluetoothNotificationsHookOptions = {
+  serviceUuid: "",
+  characteristicUuid: "",
+  deviceOptions: { acceptAllDevices: true, optionalServices: [] },
+  parser: getFloatValue,
+  onNotification: (parsed: number | string, event: Event) => {
+    return;
+  },
+  onServerDisconnect: (event: Event) => {
+    return;
+  },
+};
+
+export function useBluetoothNotifications(
+  _options: BluetoothNotificationsHookOptions = defaultHookOptions,
+): BluetoothConnections {
+  const [device, setDevice] = useState<BluetoothDevice>();
+  const [server, setServer] = useState<BluetoothRemoteGATTServer>();
+  const [service, setService] = useState<BluetoothRemoteGATTService>();
+  const [characteristic, setCharacteristic] = useState<BluetoothRemoteGATTCharacteristic>();
+  const [status, setStatus] = useState<BluetoothNotificationsStatus>(BluetoothNotificationsStatus.READY);
+
+  const {
+    serviceUuid,
+    characteristicUuid,
+    deviceOptions,
+    onNotification,
+    onServerDisconnect,
+    onError,
+    parser,
+  } = useMemo(() => {
+    return { ...defaultHookOptions, ..._options };
+  }, [_options]);
+
+  const bluetooth: Bluetooth = useMemo(() => {
+    if (!navigator.bluetooth) {
+      const error = new Error("This device is not capable of using bluetooth");
+      if (onError) {
+        onError(error);
+      } else {
+        throw error;
+      }
+    }
+    return navigator.bluetooth;
+  }, [onError]);
+
+  const handleNotifications = useCallback(
+    (event: Event) => {
+      const value = (event.target as BluetoothRemoteGATTCharacteristic).value;
+      if (value && parser) {
+        const parsed = parser(value);
+        if (onNotification) {
+          onNotification(parsed, event);
+        }
+      }
+    },
+    [onNotification, parser],
+  );
+
+  const handleError = useCallback(
+    (error: Error) => {
+      setStatus(BluetoothNotificationsStatus.ERROR);
+      if (onError) {
+        onError(error);
+      } else {
+        throw error;
+      }
+    },
+    [onError],
+  );
+
+  const reset = useCallback(() => {
+    setCharacteristic(undefined);
+    setService(undefined);
+
+    if (server) {
+      server.disconnect();
+      setServer(undefined);
+    }
+
+    setDevice(undefined);
+    setStatus(BluetoothNotificationsStatus.STOPPED);
+  }, [server]);
+
+  const startStream = useCallback(async () => {
+    try {
+      setStatus(BluetoothNotificationsStatus.STARTING);
+      let service: BluetoothRemoteGATTService;
+      let characteristic: BluetoothRemoteGATTCharacteristic;
```

Comment:
these variables are never reassigned so they could be declared as const in the `if (server)` block below